src/libstore/parsed-derivations.hh

```diff
@@ -36,6 +36,9 @@ public:
     bool willBuildLocally(Store & localStore) const;
 
     bool substitutesAllowed() const;
+
+    std::optional<nlohmann::json> prepareStructuredAttrs(std::optional<StringMap> inputRewrites, Store & store, const StorePathSet & inputPaths);
+    std::string writeStructuredAttrsShell(nlohmann::json & json);
```

Comment:
Does this method actually rely on anything other than `json`? In that case it should be a free function.